examples/silentpayments.c

```diff
@@ -0,0 +1,432 @@
+         *
+         * Here we are preventing these writes from being optimized out, as any good compiler
+         * will remove any writes that aren't used. */
+        for (i = 0; i < N_TX_INPUTS; i++) {
+            secure_erase(&sender_seckeys[i], sizeof(sender_seckeys[i]));
+        }
+    }
+
+    /*** Receiving ***/
+    {
+        /*** Transaction data ***
+         *
+         * Here we create a few global variables to represent the transaction data:
+         *
+         *     1. The transaction inputs, `tx_input_ptrs`
+         *     2. The transaction outputs, `tx_output_ptrs`
+         *
+         * These will be used to demonstrate scanning as a full node and scanning as a light client.
+         */
+        const secp256k1_xonly_pubkey *tx_input_ptrs[N_TX_INPUTS];
+        const secp256k1_xonly_pubkey *tx_output_ptrs[N_TX_OUTPUTS];
+        unsigned char light_client_data33[33];
+
+        for (i = 0; i < N_TX_INPUTS; i++) {
+            tx_input_ptrs[i] = &tx_inputs[i];
+        }
+        for (i = 0; i < N_TX_OUTPUTS; i++) {
+            tx_output_ptrs[i] = &tx_outputs[i];
+        }
+
+        {
+            /*** Scanning with labels as a full node (Bob) ***
+             *
+             * Since Bob has access to the full transaction, scanning is simple:
+             *
+             *     1. Collect the relevant public data from the transaction inputs and call
+             *        `secp256k1_silentpayments_recipient_public_data_create`
+             *     2. Call `secp256k1_silentpayments_recipient_scan_outputs`
+             *
+             * Additionally, since Bob has access to the full transaction outputs when scanning its easy for him
+             * to scan with labels, as demonstrated below. For efficient scanning, Bob keeps a cache of
+             * every label he has previously used and uses a callback to check if a potential label exists
+             * in his cache. Since the labels are created using an incremental integer `m`, if Bob ever
+             * forgets how many labels he has previously used, he can pregenerate a large number of
+             * labels e.g., 0..100_000.
+             */
+            secp256k1_silentpayments_found_output found_outputs[N_TX_OUTPUTS];
+            secp256k1_silentpayments_found_output *found_output_ptrs[N_TX_OUTPUTS];
+            secp256k1_silentpayments_public_data public_data;
+            secp256k1_pubkey spend_pubkey;
+            size_t n_found_outputs;
+            unsigned int m = 1;
+            struct labels_cache labels_cache;
+
+            for (i = 0; i < N_TX_OUTPUTS; i++) {
+                found_output_ptrs[i] = &found_outputs[i];
+            }
+
+            /* In this contrived example, our label context needs the secp256k1 context because our lookup function
+             * is calling `secp256k1_ec_pubkey_cmp`. In practice, this context can be anything the lookup function needs.
+             */
+            labels_cache.ctx = ctx;
```

Comment:
Do you think this could be simplified by working with the serializations of the pubkeys?

Reply:
Good question! Initially, I think I was worried going from `ge -> secp256k1_pubkey -> bytes` would add extra overhead during scanning, but since then I've discovered we can go from `ge -> bytes`. I also benchmarked the change and there was no difference in scanning.

I'd say this is preferable too because now the callback function doesn't need to be aware of `secp256k1_pubkey` and it's very likely the labels will be stored in the labels cache as bytes (which would mean in the previous version the callback function would need to do the serialization, anyways).

Reply:
Updated to use `unsigned char`, thanks for the suggestion!